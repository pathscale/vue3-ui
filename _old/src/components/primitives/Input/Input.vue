<script>
import { ref, watchEffect, computed } from 'vue'
import EyeIcon from './EyeIcon.vue'

export default {
  name: 'VInput',
  components: { EyeIcon },
  inheritAttrs: false,
  props: {
    color: String,
    size: String,
    rounded: Boolean,
    loading: Boolean,
    expanded: Boolean,
    modelValue: [String, Number],
    passwordReveal: Boolean,
  },
  emits: ['update:modelValue'],
  setup(props, { emit, slots, attrs }) {
    const hasLeftIcon = computed(() => Boolean(slots.leftIcon))
    const hasRightIcon = computed(() => Boolean(slots.rightIcon) || props.passwordReveal)
    const value = ref(props.modelValue)
    const showPassword = ref(false)

    const computedType = computed(() => {
      if (showPassword.value) {
        return 'text'
      }
      return attrs.type
    })

    const tooglePassword = () => {
      showPassword.value = !showPassword.value
    }

    watchEffect(() => {
      value.value = props.modelValue
    })

    watchEffect(() => emit('update:modelValue', value.value))

    return {
      value,
      hasLeftIcon,
      hasRightIcon,
      tooglePassword,
      computedType,
      rightIcon: slots.rightIcon,
      showPassword,
    }
  },
}
</script>

<template>
  <div
    class="control"
    :class="[
      size,
      {
        'is-loading': loading,
        'has-icons-left': hasLeftIcon,
        'has-icons-right': hasRightIcon,
      },
    ]">
    <input
      class="input"
      v-bind="$attrs"
      v-model="value"
      :type="computedType"
      :class="[
        color,
        size,
        {
          'is-rounded': rounded,
          'is-expanded': expanded,
        },
      ]" />
    <span class="icon is-left" v-if="hasLeftIcon">
      <slot name="leftIcon" />
    </span>
    <span class="icon is-right" v-if="hasRightIcon">
      <div class="is-clickable" v-if="passwordReveal" @click="tooglePassword">
        <slot v-if="rightIcon" name="rightIcon" />
        <div v-else>
          <eye-icon :invisible="showPassword" />
        </div>
      </div>
      <slot v-else name="rightIcon" />
    </span>
  </div>
</template>
