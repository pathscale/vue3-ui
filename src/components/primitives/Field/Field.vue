<script>
import { computed } from 'vue'

export default {
  name: 'VField',
  props: {
    type: String,
    label: String,
    labelFor: String,
    message: String,
    grouped: Boolean,
    groupMultiline: Boolean,
    position: String,
    expanded: Boolean,
    horizontal: Boolean,
    addons: Boolean,
    size: String,
  },
  setup(props) {
    const hasHorizontalLabel = computed(() => props.horizontal && props.label)
    return { hasHorizontalLabel }
  }
}
</script>

<template>
  <div class="field" :class="[
    position, {
      'is-expanded': expanded,
      'is-horizontal': horizontal,
      'is-grouped': grouped,
      'is-grouped-multiline': groupMultiline,
      'has-addons': addons
    }]">
    <div v-if="hasHorizontalLabel" class="field-label" :class="size">
      <label class="label" :for="labelFor">{{ label }}</label>
    </div>
    <label v-else-if="label" class="label" :for="labelFor">{{ label }}</label>
    <div v-if="horizontal" class="field-body">
      <div class="field" :class="{'is-grouped': grouped}">
        <slot :rounded="type" />
      </div>
    </div>
    <slot :color="type" v-else />
    <p v-if="message" class="help" :class="type">
      {{ message }}
    </p>
  </div>
</template>
