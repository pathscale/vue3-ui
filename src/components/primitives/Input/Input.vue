<script setup lang="ts">
import { computed, ref, useAttrs, useSlots } from "vue";
import EyeIcon from "./EyeIcon.vue";

const props = defineProps<{
  color?: string;
  size?: string;
  rounded?: boolean;
  loading?: boolean;
  expanded?: boolean;
  passwordReveal?: boolean;
}>();

const value = defineModel();

const slots = useSlots();
const attrs = useAttrs();

const hasLeftIcon = computed(() => Boolean(slots.leftIcon));
const hasRightIcon = computed(
  () => Boolean(slots.rightIcon) || props.passwordReveal,
);

const showPassword = ref(false);

const computedType = computed(() => {
  if (showPassword.value) {
    return "text";
  }
  return attrs.type;
});

const tooglePassword = () => {
  showPassword.value = !showPassword.value;
};
</script>

<template>
  <div
    class="control"
    :class="[
      size,
      {
        'is-loading': loading,
        'has-icons-left': hasLeftIcon,
        'has-icons-right': hasRightIcon,
      },
    ]">
    <input
      class="input"
      v-bind="$attrs"
      v-model="value"
      :type="computedType"
      :class="[
        color,
        size,
        {
          'is-rounded': rounded,
          'is-expanded': expanded,
        },
      ]" />
    <span class="icon is-left" v-if="hasLeftIcon">
      <slot name="leftIcon" />
    </span>
    <span class="icon is-right" v-if="hasRightIcon">
      <div class="is-clickable" v-if="passwordReveal" @click="tooglePassword">
        <slot v-if="$slots.rightIcon" name="rightIcon" />
        <div v-else>
          <eye-icon :invisible="showPassword" />
        </div>
      </div>
      <slot v-else name="rightIcon" />
    </span>
  </div>
</template>
