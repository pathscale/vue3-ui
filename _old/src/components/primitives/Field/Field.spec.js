import { shallowMount, mount } from '@vue/test-utils'
import Field from './Field.vue'
import Input from '../Input/Input.vue'

let wrapper

describe('Field', () => {
  beforeEach(() => {
    wrapper = shallowMount(Field)
  })

  it('is called', () => {
    expect(wrapper.exists()).toBeTruthy()
  })

  it('render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('adds a help <p> element in the root div.field when "message" prop is passed', async () => {
    wrapper = mount(Field, {
      slots: {
        default: [Input],
      },
      props: {
        type: 'is-danger',
      },
    })

    const message = 'This is a message'

    expect(wrapper.findAll('p')).toEqual([])

    wrapper.setProps({ message })
    await wrapper.vm.$nextTick()
    expect(wrapper.find('p').classes()).toEqual(['help', 'is-danger'])
    expect(wrapper.find('p').text()).toBe(message)
  })

  it('adds a label element under the root div.field when "label" prop is passed', () => {
    wrapper = shallowMount(Field, {
      propsData: {
        label: 'Some label',
      },
    })
    expect(wrapper.find('label').text()).toBe('Some label')
  })
})
