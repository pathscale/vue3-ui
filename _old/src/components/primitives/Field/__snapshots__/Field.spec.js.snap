// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field render correctly 1`] = `
<div class="field">
  <!--v-if-->
  <!--v-if-->
</div>
`;
